// typescript-to-lua/dist/transpilation/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.transpileString = exports.transpileVirtualProject = exports.createVirtualProgram = exports.transpileProject = exports.transpileFiles = void 0;
const fs = require("fs");
const path = require("path");
const ts = require("typescript");
const tsconfig_1 = require("../cli/tsconfig");
const emit_1 = require("./emit");
const transpile_1 = require("./transpile");
__exportStar(require("./emit"), exports);
__exportStar(require("./transpile"), exports);
function transpileFiles(rootNames, options = {}) {
    const program = ts.createProgram(rootNames, options);
    const { transpiledFiles, diagnostics: transpileDiagnostics } = transpile_1.transpile({ program });
    const emitResult = emit_1.emitTranspiledFiles(program, transpiledFiles);
    const diagnostics = ts.sortAndDeduplicateDiagnostics([
        ...ts.getPreEmitDiagnostics(program),
        ...transpileDiagnostics,
    ]);
    return { diagnostics: [...diagnostics], emitResult };
}
exports.transpileFiles = transpileFiles;
function transpileProject(configFileName, optionsToExtend) {
    const parseResult = tsconfig_1.parseConfigFileWithSystem(configFileName, optionsToExtend);
    if (parseResult.errors.length > 0) {
        return { diagnostics: parseResult.errors, emitResult: [] };
    }
    return transpileFiles(parseResult.fileNames, parseResult.options);
}
exports.transpileProject = transpileProject;
const libCache = {};
/** @internal */
function createVirtualProgram(input, options = {}) {
    const compilerHost = {
        fileExists: () => true,
        getCanonicalFileName: fileName => fileName,
        getCurrentDirectory: () => "",
        getDefaultLibFileName: ts.getDefaultLibFileName,
        readFile: () => "",
        getNewLine: () => "\n",
        useCaseSensitiveFileNames: () => false,
        writeFile() { },
        getSourceFile(filename) {
            if (filename in input) {
                return ts.createSourceFile(filename, input[filename], ts.ScriptTarget.Latest, false);
            }
            if (filename.startsWith("lib.")) {
                if (libCache[filename])
                    return libCache[filename];
                const typeScriptDir = path.dirname(require.resolve("typescript"));
                const filePath = path.join(typeScriptDir, filename);
                const content = fs.readFileSync(filePath, "utf8");
                libCache[filename] = ts.createSourceFile(filename, content, ts.ScriptTarget.Latest, false);
                return libCache[filename];
            }
        },
    };
    return ts.createProgram(Object.keys(input), options, compilerHost);
}
exports.createVirtualProgram = createVirtualProgram;
function transpileVirtualProject(files, options = {}) {
    const program = createVirtualProgram(files, options);
    const result = transpile_1.transpile({ program });
    const diagnostics = ts.sortAndDeduplicateDiagnostics([...ts.getPreEmitDiagnostics(program), ...result.diagnostics]);
    return { ...result, diagnostics: [...diagnostics] };
}
exports.transpileVirtualProject = transpileVirtualProject;
function transpileString(main, options = {}) {
    const { diagnostics, transpiledFiles } = transpileVirtualProject({ "main.ts": main }, options);
    return { diagnostics, file: transpiledFiles.find(({ fileName }) => fileName === "main.ts") };
}
exports.transpileString = transpileString;
