// @hydro-sdk/hydro-sdk/runtime/flutter/rendering/boxConstraints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BoxConstraints = void 0;
const double_1 = require("./../../dart/core/double");
const type_1 = require("./../../dart/core/type");
const syntheticBox_1 = require("./../../syntheticBox");
class BoxConstraints extends syntheticBox_1.JITAllocatingRTManagedBox {
    constructor(props) {
        super();
        this.internalRuntimeType = new type_1.Type(BoxConstraints);
        this.props = undefined;
        this.minWidth = props.minWidth;
        this.maxWidth = props.maxWidth;
        this.minHeight = props.minHeight;
        this.maxHeight = props.maxHeight;
        if (this.minWidth === undefined) {
            this.minWidth = 0.0;
        }
        if (this.maxWidth === undefined) {
            this.maxWidth = double_1.double.infinity;
        }
        if (this.minHeight === undefined) {
            this.minHeight = 0.0;
        }
        if (this.maxHeight === undefined) {
            this.maxHeight = double_1.double.infinity;
        }
    }
    unwrap() {
        return flutter.rendering.boxConstraints({
            minWidth: this.minWidth,
            maxWidth: this.maxWidth,
            minHeight: this.minHeight,
            maxHeight: this.maxHeight,
        });
    }
}
exports.BoxConstraints = BoxConstraints;
