// @hydro-sdk/hydro-sdk/runtime/flutter/services/textSelection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextSelection = void 0;
const type_1 = require("../../dart/core/type");
const textAffinity_1 = require("../../dart/ui/textAffinity");
const syntheticBox_1 = require("../../syntheticBox");
class TextSelection extends syntheticBox_1.JITAllocatingRTManagedBox {
    constructor(props) {
        super();
        this.internalRuntimeType = new type_1.Type(TextSelection);
        this.props = props;
        if (this.props.affinity === undefined) {
            this.props.affinity = textAffinity_1.TextAffinity.downstream;
        }
        if (this.props.isDirectional === undefined) {
            this.props.isDirectional = false;
        }
    }
    static collapsed(props) {
        return new TextSelection({
            baseOffset: props.offset,
            extentOffset: props.offset,
            isDirectional: false,
            affinity: props.affinity,
        });
    }
    //@todo TextSelection#fromPosition : (position : TextPosition) => TextSelection
    unwrap() {
        return flutter.services.textSelection(this.props);
    }
}
exports.TextSelection = TextSelection;
