// @hydro-sdk/hydro-sdk/runtime/flutter/services/textInputAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TextInputAction = void 0;
var TextInputAction;
(function (TextInputAction) {
    TextInputAction[TextInputAction["none"] = 0] = "none";
    TextInputAction[TextInputAction["unspecified"] = 1] = "unspecified";
    TextInputAction[TextInputAction["done"] = 2] = "done";
    TextInputAction[TextInputAction["go"] = 3] = "go";
    TextInputAction[TextInputAction["search"] = 4] = "search";
    TextInputAction[TextInputAction["send"] = 5] = "send";
    TextInputAction[TextInputAction["next"] = 6] = "next";
    TextInputAction[TextInputAction["previous"] = 7] = "previous";
    TextInputAction[TextInputAction["continueAction"] = 8] = "continueAction";
    TextInputAction[TextInputAction["join"] = 9] = "join";
    TextInputAction[TextInputAction["route"] = 10] = "route";
    TextInputAction[TextInputAction["emergencyCall"] = 11] = "emergencyCall";
    TextInputAction[TextInputAction["newline"] = 12] = "newline";
})(TextInputAction = exports.TextInputAction || (exports.TextInputAction = {}));
