// @hydro-sdk/hydro-sdk/runtime/flutter/widgets/image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Image = void 0;
const type_1 = require("../../dart/core/type");
const filterQuality_1 = require("../../dart/ui/filterQuality");
const alignment_1 = require("../painting/alignment");
const imageRepeat_1 = require("../painting/imageRepeat");
const networkImage_1 = require("../painting/networkImage");
const resizeImage_1 = require("../painting/resizeImage");
const statelessWidget_1 = require("./statelessWidget");
class Image extends statelessWidget_1.StatelessWidget {
    constructor(src, props) {
        super();
        this.internalRuntimeType = new type_1.Type(Image);
        this.src = src;
        this.props = props;
    }
    static network(src, props) {
        const { key, scale = 1.0, semanticLabel, excludeFromSemantics = false, width, height, color, colorBlendMode, fit, alignment = alignment_1.Alignment.center, repeat = imageRepeat_1.ImageRepeat.noRepeat, matchTextDirection = false, gaplessPlayback = false, filterQuality = filterQuality_1.FilterQuality.low, cacheWidth, cacheHeight, } = props;
        return new Image(src, {
            key,
            scale,
            semanticLabel,
            excludeFromSemantics,
            width,
            height,
            color,
            colorBlendMode,
            fit,
            alignment,
            repeat,
            matchTextDirection,
            gaplessPlayback,
            filterQuality,
            image: resizeImage_1.ResizeImage.resizeIfNeeded(cacheWidth, cacheHeight, new networkImage_1.NetworkImage(src, { scale: scale })),
        });
    }
    build() {
        return flutter.widgets.image(this.props);
    }
}
exports.Image = Image;
