// @hydro-sdk/hydro-sdk/runtime/flutter/widgets/positioned.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Positioned = void 0;
const type_1 = require("../../dart/core/type");
const textDirection_1 = require("../../dart/ui/textDirection");
const statelessWidget_1 = require("./statelessWidget");
class Positioned extends statelessWidget_1.StatelessWidget {
    constructor(props) {
        super();
        this.internalRuntimeType = new type_1.Type(Positioned);
        this.props = props;
    }
    static directional(props) {
        let left;
        let right;
        switch (props.textDirection) {
            case textDirection_1.TextDirection.rtl:
                left = props.end;
                right = props.start;
                break;
            case textDirection_1.TextDirection.ltr:
                left = props.start;
                right = props.end;
                break;
            default:
                ((args) => {
                    throw new Error("");
                })(props.textDirection);
                break;
        }
        return new Positioned({
            key: props.key,
            left,
            top: props.top,
            right,
            bottom: props.bottom,
            width: props.width,
            height: props.height,
            child: props.child,
        });
    }
    static fromRect(props) {
        return new Positioned({
            left: props.rect.left,
            top: props.rect.top,
            right: props.rect.right,
            bottom: props.rect.bottom,
            key: props.key,
            child: props.child,
        });
    }
    static fill(props) {
        const { left = 0.0, top = 0.0, right = 0.0, bottom = 0.0 } = props;
        return new Positioned({
            key: props.key,
            child: props.child,
            left,
            top,
            right,
            bottom,
        });
    }
    build() {
        return flutter.widgets.positioned(this.props);
    }
}
exports.Positioned = Positioned;
