// @hydro-sdk/hydro-sdk/runtime/dart/ui/blendMode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlendMode = void 0;
var BlendMode;
(function (BlendMode) {
    BlendMode[BlendMode["clear"] = 0] = "clear";
    BlendMode[BlendMode["src"] = 1] = "src";
    BlendMode[BlendMode["dst"] = 2] = "dst";
    BlendMode[BlendMode["srcOver"] = 3] = "srcOver";
    BlendMode[BlendMode["dstOver"] = 4] = "dstOver";
    BlendMode[BlendMode["srcIn"] = 5] = "srcIn";
    BlendMode[BlendMode["dstIn"] = 6] = "dstIn";
    BlendMode[BlendMode["srcOut"] = 7] = "srcOut";
    BlendMode[BlendMode["dstOut"] = 8] = "dstOut";
    BlendMode[BlendMode["srcATop"] = 9] = "srcATop";
    BlendMode[BlendMode["dstATop"] = 10] = "dstATop";
    BlendMode[BlendMode["xor"] = 11] = "xor";
    BlendMode[BlendMode["plus"] = 12] = "plus";
    BlendMode[BlendMode["modulate"] = 13] = "modulate";
    BlendMode[BlendMode["screen"] = 14] = "screen";
    BlendMode[BlendMode["overlay"] = 15] = "overlay";
    BlendMode[BlendMode["darken"] = 16] = "darken";
    BlendMode[BlendMode["lighten"] = 17] = "lighten";
    BlendMode[BlendMode["colorDodge"] = 18] = "colorDodge";
    BlendMode[BlendMode["colorBurn"] = 19] = "colorBurn";
    BlendMode[BlendMode["hardLight"] = 20] = "hardLight";
    BlendMode[BlendMode["softLight"] = 21] = "softLight";
    BlendMode[BlendMode["difference"] = 22] = "difference";
    BlendMode[BlendMode["exclusion"] = 23] = "exclusion";
    BlendMode[BlendMode["multiply"] = 24] = "multiply";
    BlendMode[BlendMode["hue"] = 25] = "hue";
    BlendMode[BlendMode["saturation"] = 26] = "saturation";
    BlendMode[BlendMode["color"] = 27] = "color";
    BlendMode[BlendMode["luminosity"] = 28] = "luminosity";
})(BlendMode = exports.BlendMode || (exports.BlendMode = {}));
