// @hydro-sdk/hydro-sdk/runtime/flutter/painting/alignment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Alignment = void 0;
const syntheticBox_1 = require("./../../syntheticBox");
const type_1 = require("../../dart/core/type");
class Alignment extends syntheticBox_1.JITAllocatingRTManagedBox {
    constructor(x, y) {
        super();
        this.internalRuntimeType = new type_1.Type(Alignment);
        this.props = undefined;
        this.x = x;
        this.y = y;
    }
    unwrap() {
        return flutter.painting.alignment(this.x, this.y);
    }
}
exports.Alignment = Alignment;
Alignment.topLeft = new Alignment(-1.0, -1.0);
Alignment.topCenter = new Alignment(0.0, -1.0);
Alignment.topRight = new Alignment(1.0, -1.0);
Alignment.centerLeft = new Alignment(-1.0, 0.0);
Alignment.center = new Alignment(0.0, 0.0);
Alignment.centerRight = new Alignment(1.0, 0.0);
Alignment.bottomLeft = new Alignment(-1.0, 1.0);
Alignment.bottomCenter = new Alignment(0.0, 1.0);
Alignment.bottomRight = new Alignment(1.0, 1.0);
