// @hydro-sdk/hydro-sdk/runtime/flutter/painting/borderSide.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BorderSide = void 0;
const color_1 = require("./../../dart/ui/color");
const type_1 = require("../../dart/core/type");
const syntheticBox_1 = require("../../syntheticBox");
const borderStyle_1 = require("./borderStyle");
class BorderSide extends syntheticBox_1.JITAllocatingRTManagedBox {
    constructor(props) {
        super();
        this.internalRuntimeType = new type_1.Type(BorderSide);
        this.props = props;
        if (this.props.color === undefined) {
            this.props.color = new color_1.Color(0xff000000);
        }
        if (this.props.width === undefined) {
            this.props.width = 1.0;
        }
        if (this.props.style === undefined) {
            this.props.style = borderStyle_1.BorderStyle.solid;
        }
    }
    unwrap() {
        return flutter.painting.borderSide(this.props);
    }
}
exports.BorderSide = BorderSide;
BorderSide.none = new BorderSide({
    width: 0.0,
    style: borderStyle_1.BorderStyle.none,
});
