// @hydro-sdk/hydro-sdk/runtime/flutter/painting/borderRadius.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BorderRadius = void 0;
const type_1 = require("../../dart/core/type");
const radius_1 = require("../../dart/ui/radius");
class BorderRadius {
    constructor() {
        this.internalRuntimeType = new type_1.Type(BorderRadius);
    }
    static all(radius) {
        return BorderRadius.only({
            topLeft: radius,
            topRight: radius,
            bottomLeft: radius,
            bottomRight: radius,
        });
    }
    static circular(radius) {
        return BorderRadius.all(radius_1.Radius.circular(radius));
    }
    static vertical(props) {
        const { top = radius_1.Radius.zero, bottom = radius_1.Radius.zero } = props;
        return BorderRadius.only({
            topLeft: top,
            topRight: top,
            bottomLeft: bottom,
            bottomRight: bottom,
        });
    }
    static horizontal(props) {
        const { left = radius_1.Radius.zero, right = radius_1.Radius.zero } = props;
        return BorderRadius.only({
            topLeft: left,
            topRight: right,
            bottomLeft: left,
            bottomRight: right,
        });
    }
    static only(props) {
        const { topLeft = radius_1.Radius.zero, topRight = radius_1.Radius.zero, bottomLeft = radius_1.Radius.zero, bottomRight = radius_1.Radius.zero, } = props;
        return flutter.painting.borderRadiusOnly({
            topLeft,
            topRight,
            bottomLeft,
            bottomRight,
        });
    }
}
exports.BorderRadius = BorderRadius;
BorderRadius.zero = BorderRadius.all(radius_1.Radius.zero);
