// @hydro-sdk/hydro-sdk/runtime/scopedModel/scopedModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopedModel = void 0;
const animatedBuilder_1 = require("../flutter/widgets/animatedBuilder");
const statelessWidget_1 = require("../flutter/widgets/statelessWidget");
const inheritedModel_1 = require("./inheritedModel");
class ScopedModel extends statelessWidget_1.StatelessWidget {
    constructor({ model, child }) {
        super();
        this.model = model;
        this.child = child;
    }
    build() {
        return new animatedBuilder_1.AnimatedBuilder({
            animation: this.model,
            builder: (context) => new inheritedModel_1.InheritedModel({
                model: this.model,
                child: this.child,
            }),
        });
    }
    static of(context, type) {
        const inheritedModel = context.ancestorInheritedElementForWidgetOfExactType(type);
        return inheritedModel === null || inheritedModel === void 0 ? void 0 : inheritedModel.model;
    }
}
exports.ScopedModel = ScopedModel;
